src/azul/plugins/repository/tdr/__init__.py

```diff
@@ -240,6 +262,124 @@ def verify_authorization(self):
+                require(associate_type == 'project',
+                        f'Supplementary file must be associated with entity of type "project", '
+                        f'not "{associate_type}"')
+                require(associate_id == bundle_project_id,
+                        f'Supplementary file must be associated with the current project '
+                        f'({bundle_project_id}, not {associate_id})')
+                for supp_file in cast(JSONs, link['files']):
+                    entities['supplementary_file'].add(supp_file['file_id'])
+            else:
+                raise RequirementError(f'Unexpected link_type: {link_type}')
+        return entities, inputs, outputs
+
+    def _find_dangling_inputs(self,
+                              inputs: TypedEntities,
+                              outputs: Collection[EntityID]
+                              ) -> TypeGroupedEntities:
+        """
+        :param inputs: Mapping of input entity ID's to entity types.
+        :param outputs: Collection of output entity ID's.
+        :return: Input entities are files and do not occur in `outputs`, grouped by entity type.
+        """
+        grouped = defaultdict(set)
+        for input_id, input_type in inputs.items():
+            if input_type.endswith('_file') and input_id not in outputs:
+                grouped[input_type].add(input_id)
+        return grouped
+
+    def _link_file_ids(self,
+                       input_type: EntityType,
+                       input_ids: Set[EntityID]
+                       ) -> EntityMapping:
+        """
+        File entities are represented inconstently when they appear as inputs
+        versus as outputs.
+        Returns a mapping of output ID's (which use the metadata key column) to
+        input ID's (which use the data descriptor `file_id`).
+        """
+        file_id = 'JSON_EXTRACT_SCALAR(descriptor, "$.file_id")'
+        rows = self._run_sql(f'''
+            SELECT {input_type}_id AS output_id, {file_id} AS input_id
+            FROM {self._source.bq_name}.{input_type}
+            WHERE {file_id} IN ({selection_literal(map(repr, input_ids))})
+        ''')
+        outputs_to_inputs = {row['output_id']: row['input_id'] for row in rows}
+        missing = input_ids - set(outputs_to_inputs.values())
+        require(not missing,
+                f'Failed to locate input entities in {self._source.bq_name}.{input_type}: {missing}')
+        return outputs_to_inputs
+
+    def _find_upstream_bundles(self,
+                               outputs: Set[EntityID]
+                               ) -> Set[BundleFQID]:
+        """
+        Search for bundles containing the specified output entities.
+        """
+
+        outputs_literal = selection_literal(map(repr, outputs))
+        # This query is incompatible with TinyQuery due its extensive usage of
+        # BigQuery arrays.
+        # Not sure how to approach this for unit testing.
```

Comment:
The best I can come up with is to extract this _run_sql call into a separate method and mock it during the test.

Maybe we should ditch TinyQuery and simply mock _run_sql to return canned responses. We'd simply can a sequence of return values in a list. The first call to _run_sql mock would return the first value in the list, the second call the second one and so on.

Reply:
I forgot that we added a subclass of the plugin for testing. That makes it easy to work around this by overriding this particular method.